src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java

```diff
@@ -0,0 +1,304 @@
+        fetchSchema(client);
+      }
+    } catch (Exception e) {
+      results = Collections.emptyList();
+      LOG.error("Error while fetching data", e);
+    }
+
+    iterator = results.iterator();
+  }
+
+  private void fetchSchema(MarketingCloudClient client) {
+    //Fetch the column definition
+    List<Schema.Field> schemaFields;
+
+    try {
+      if (object == SourceObject.DATA_EXTENSION) {
+        sfObjectMetaData = client.fetchDataExtensionSchema(dataExtensionKey);
+      } else {
+        sfObjectMetaData = client.fetchObjectSchema(object);
+      }
+
+      //Build schema
+      tableFields = sfObjectMetaData.getSchema().getFields();
+      schemaFields = new ArrayList<>(tableFields);
+
+      if (pluginConf.getQueryMode() == SourceQueryMode.MULTI_OBJECT) {
+        schemaFields.add(Schema.Field.of(tableNameField, Schema.of(Schema.Type.STRING)));
+      }
+    } catch (Exception e) {
+      schemaFields = Collections.emptyList();
+    }
+
+    schema = Schema.recordOf(tableName.replaceAll("-", "_"), schemaFields);
+  }
+
+  /**
+   * Read data from ETApiObject and convert it to StructureRecord.
+   */
+  private void convertRecord(StructuredRecord.Builder recordBuilder, ETApiObject row) {
+    for (Schema.Field field : tableFields) {
+      String fieldName = field.getName();
+      Object rawFieldValue = null;
+      if (row instanceof ETDataExtensionRow) {
+        String apiFieldName = sfObjectMetaData.lookupFieldsMap(fieldName);
+        rawFieldValue = ((ETDataExtensionRow) row).getColumn(apiFieldName);
+      } else {
+        rawFieldValue = getFieldValue(row, fieldName);
+      }
+
+      Object fieldValue = convertToValue(fieldName, field.getSchema(), rawFieldValue);
+
+      recordBuilder.set(fieldName, fieldValue);
+    }
+  }
+
+  /**
+   * Read data from ETApiObject using reflection for a given field name.
+   */
+  private Object getFieldValue(ETApiObject row, String fieldName) {
+    try {
+      Method method = row.getClass().getMethod(createGetterName(fieldName));
+      return method.invoke(row);
+    } catch (Exception e) {
+      LOG.error(String.format("Error while fetching %s.%s value", row.getClass().getSimpleName(), fieldName), e);
+      return null;
+    }
+  }
+
+  /**
+   * Constructs the get method name to be used in reflection call.
+   */
+  private String createGetterName(String name) {
+    StringBuilder sb = new StringBuilder("get");
+    sb.append(name.substring(0, 1).toUpperCase());
+    sb.append(name.substring(1));
+    return sb.toString();
+  }
+
+  /**
+   * Converts raw field value according to the schema field type.
+   */
+  private Object convertToValue(String fieldName, Schema fieldSchema, Object fieldValue) {
+    Schema.Type fieldType = fieldSchema.getType();
+    Schema.LogicalType logicalType = fieldSchema.getLogicalType();
+    if (fieldSchema.getLogicalType() != null) {
+      return transformLogicalType(fieldName, logicalType, fieldValue);
+    }
+
+    switch (fieldType) {
```

Comment:
Seems like not all types are handled. If the types are not supported by salesforce then a clear documentation on what types are supported and how they are mapped should be documented.

Reply:
Its documented in the design doc.

Reply:
should be documented in the plugin doc

Reply:
I will update the Documentation.